### DISCUSSION TITLE
How to identify nodes inside a Morphir IR

### DISCUSSION THREAD

**Discussion by bekand** [finos/morphir#145] [Ideas]
## Introduction
Several requirements of the Decorations feature require us to have the ability to refer to specific nodes insode of a Morphir IR. 

## User stories

* As a Morphir modeler I would like to have a way to uniquely identify nodes in the Morphir IR for the purposes of decorating them.
	1. I would like any such identifier to be able to modules, top level types and values, or types / values inside other types / values.
	2. Additionally, I would like to be able to refer to function arguments and record fields in a convenient manner, since these nodes are common candidates for decoration.
	3. Finally, I would like any such identifier to have a well defined structure.

## Resulting questions

1. What is something that uniquely identifies a Morphir module, type or value?
2. How do we refer to a node inside  a top level type or value? What is something that uniquely identifies these nodes?
3. How do we make it as intuitive and convenient as possible to refer to concepts which are frequent candidates for decoration?
4. What should the structure of the resulting identifier be?

**Comment by DamianReeves** [comment]
This sounds a lot like the concept of [ShapeId](https://smithy.io/2.0/spec/model.html#grammar-token-smithy-ShapeId) in smithy.

I thought for Morphir it might make sense to have an equivalent thing known as `ConceptId`.

**Reply by stephengoldbaum** [reply]
Yeah, that's the same idea.  Is there no pre-existing URI standard for classes, types, functions, members, etc.?  Like from OMG or something?
